src/leiningen/ring/uberwar.clj

```diff
@@ -13,25 +17,54 @@
     (get-cp project)
     (->> (:library-path project) io/file .listFiles (map str))))
 
-(defn contains-entry? [^java.io.File file ^String entry]
-  (with-open [jar-file (JarFile. file)]
-    (some (partial = entry)
-          (map #(.getName ^JarEntry %)
-               (enumeration-seq (.entries jar-file))))))
-
 (defn jar-dependencies [project]
   (for [pathname (get-classpath project)
-        :let [file (io/file pathname)
-              fname (.getName file)]
-        :when (and (.endsWith fname ".jar")
-                   ;; Servlet container will have it's own servlet-api impl
-                   (not (contains-entry? file "javax/servlet/Servlet.class")))]
-    file))
-
-(defn jar-entries [war project]
+        :when (str/ends-with? pathname ".jar")]
+    (io/file pathname)))
+
+(def ^:private javax-servlet? #{"javax/servlet/Servlet.class"})
+
+(defn- pom-properties? [file-name]
+  (str/ends-with? file-name "pom.properties"))
+
+(defn- read-jar-pom-properties [^String full-path]
+  (with-open [in (.openStream (URL. full-path))]
+    (->> (doto (Properties.) (.load in))
+         (into {}))))
+
+(defn- from-pom-properties
+  "Returns a jar name which includes the groupId, artifactId, & version,
+   as read by the pom.properties file inside the jar file found at <jar-path>."
+  [jar-path pom-properties-jar-path]
+  (let [pom-properties-full-path (str "jar:file:" jar-path "!/" pom-properties-jar-path)
+        {:strs [groupId artifactId version]} (read-jar-pom-properties pom-properties-full-path)]
+    (when (and groupId artifactId version)
+      ;; found pom.properties - reconstruct the jar name to avoid collisions (issue #216)
+      (str groupId \- artifactId \- version ".jar"))))
+
+(defn- war-path-for-jar [^File jar]
+  (with-open [jar-file (JarFile. jar)]
+    (let [javax&pom (->> (.entries jar-file)
+                         enumeration-seq
+                         (map (fn [^JarEntry je] (.getName je)))
+                         (filter (some-fn javax-servlet? pom-properties?)))] ;; 2 elements maximum
```

Comment:
Why 2 elements maximum? What is this comment conveying? This threading could also be split into a function that would explain what it does.

Reply:
Do you want me to remove the comment or expand on it?

Reply:
I don't understand why you're searching for both the servlet class and the pom.properties file at the same time. Why not just compile a list of entries, then do a search for one, then a search for the other?

```clojure
(defn- war-path-for-jar [^File jar]
  (with-open [jar-file (JarFile. jar)]
    (let [entries (find-jar-entry-names jar-file)]
      (when-not (some #{javax-servlet-class} entries)
        (path-from-pom-properties (find-pom-properties entries))))))
```

Or more abstractly:

```clojure
(defn- war-path-for-jar [^File jar]
  (with-open [jar-file (JarFile. jar)]
    (when-not (contains-javax-servlet-class? jar-file)
      (path-from-pom-properties (read-pom-properties jar-file)))))
```

Reply:
Searching for both in one-pass, is an optimisation, simply because JARs can (and often do) contain thousands of files. That is partly why the `uberwar` task takes so long. Repeating the search is quite wasteful i think. If you insist, I will split it, but please consider the implications. This search (and compiling) make up the majority of the time spent.

Reply:
In general, we should optimise first for readability and then for performance, if necessary. Readability is often more important, but it is also easier to evaluate at a glance. Performance can be counter-intuitive, and generally you want to be running benchmarks as you go to to check your assumptions.

Since we're looking for an exact match in the case of looking for the servlet class, my first thought was to check to see if the `JarFile` class has an index lookup and then benchmark that. I ran some tests with a jruby jar I had lying around, as that has 13,000 entries:

```clojure
user=> (quick-bench (.getEntry j "org/jruby/util/io/EncodingUtils.class"))
Execution time mean : 766.012522 ns

user=> (quick-bench (some #{"org/jruby/util/io/EncodingUtils.class"} (map (fn [^java.util.jar.JarEntry je] (.getName je)) (enumeration-seq (.entries j)))))
Execution time mean : 3.253957 ms
```

You can see `.getEntry` is around 4000x faster. So if we want to optimize checks for the servlet class, then we'd write a function like:

```clojure
(defn- contains-javax-servlet-class? [^JarFile jar-file]
  (some? (.getEntry jar-file "javax/servlet/Servlet.class")))
```

Performance should start with careful reading of the classes and methods being used, and then benchmarked to confirm our assumptions.

Reply:
I won't argue with anything you said, but I will say that  this is how it should have been implemented in the first place. I was only trying to make the least amount of (potentially breaking) changes. I wasn't really trying to make things faster in the code I added - I was merely trying to **not** make them slower. I will refactor this first thing tomorrow morning - thanks 👍

Reply:
This is done